### DISCUSSION TITLE
Vector type needed for storing OpenAI embeddings

### DISCUSSION THREAD

**Discussion by gablabelle** [prisma/prisma#18220] [Q&A] [answered]
Hello,

Still a Prisma newbie and don't know a great deal about DBs TBH. I was wondering if it's possible create a column of the [Vector type](https://github.com/pgvector/pgvector) needed for storing OpenAI embeddings as described in this [Supabase blog post](https://supabase.com/blog/openai-embeddings-postgres-vector#how-do-embeddings-work) but using my current [neon database](https://neon.tech/) that supports [pgvector](https://github.com/pgvector/pgvector).

So I've added the extension to my [neon database](https://neon.tech/).

```sql
create extension vector;
```

And now would need to do the equivalent of the following via the Prisma Schema:

```sql
create table documents (
  id bigserial primary key,
  content text,
  embedding vector (1536)
);
```

<img width="484" alt="Screenshot 2023-03-05 at 19 58 50" src="https://user-images.githubusercontent.com/1565793/222997159-5a94f438-ff67-488f-ba97-a01097147e2c.png">

So my guess is that this Vector type isn't supported? What happens then if I have a type supported by the DB (PostgreSQL) but not by Prisma? How do I keep all that in sync?

**Comment by janpio** [comment, marked as answer]
This should help:

> The Unsupported attribute allows you to define fields in the Prisma schema for database types that are not yet supported by Prisma. For example, MySQL's POLYGON type is not currently supported by Prisma, but can now be added to the Prisma schema using the Unsupported("polygon") type.

https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#:~:text=The%20Unsupported%20attribute%20allows%20you,(%22polygon%22)%20type.

**Reply by pranavp10** [reply]
` embedding  Unsupported("vector")?` is not working with neon databse

**Reply by pranavp10** [reply]
<img width="641" alt="image" src="https://github.com/prisma/prisma/assets/49075129/0b31146e-e033-44cc-bb24-3d21d80431db">

I am getting this error

**Reply by janpio** [reply]
The error you are getting is missing.

**Reply by pranavp10** [reply]
got the answer (https://github.com/prisma/prisma/discussions/18220#discussioncomment-6013915)

**Comment by pranavp10** [comment]
it should be `embedding  Unsupported("vector(1536)")?`

**Comment by gablabelle** [comment]
There is a complete prisma schema example using vector available here: https://github.com/vercel/examples/blob/main/storage/postgres-pgvector/prisma/schema.prisma

**Reply by pranavp10** [reply]
works thanks

**Reply by tholmes59** [reply]
This link no longer is active. Is there another resource that provides a solution that this has moved to?

**Reply by NickFoden** [reply]
Maybe for ideas check out 

https://js.langchain.com/v0.2/docs/integrations/vectorstores/prisma/

(I'm researching / looking around as well)

**Reply by tholmes59** [reply]
Thanks for passing this along!

**Reply by axel-blade** [reply]
Updated link:
https://github.com/vercel/examples/blob/nuxt-blob-example/storage/postgres-pgvector/prisma/schema.prisma

**Comment by gaufranc** [comment]
hey @pranavp10 did you get this working?

if i do the following:
const vector = await prisma.$executeRaw`SELECT "mycolumn"::vector FROM "mytable" WHERE id = ${id}`;

i always get 1 as response (1 as type number) no matter if i already saved an embedding inside

**Reply by disbelief** [reply]
What you need to do is cast the column as `::text` and then parse it into a vector (array of floats) yourself:

```typescript
const result = await client.$queryRawUnsafe<{ mycolumn: string | null }[]>(`
  SELECT "mycolumn"::text FROM "mytable"
  WHERE "id" = '${id}'
  LIMIT 1;
`);
const embeddingString = result[0]?.mycolumn;
let embedding: number[] | null = null;
if (embeddingString !== null || typeof embeddingString !== 'undefined') {
  embedding = embeddingString
    .slice(1, embedding.length - 1) // remove the brackets []
    .split(',')
    .map((v) => parseFloat(v));
}
```

**Comment by davidebukali** [comment]
This migration with a vector column works for me after adding `previewFeatures = ["postgresqlExtensions"]` to generator client and `extensions = [vector]` to datasource db.

<pre>
      generator client {
        provider = "prisma-client-js"
        previewFeatures = ["postgresqlExtensions"]
      }
</pre>

<pre>
      datasource db {
        provider     = "postgresql"
        url          = env("DATABASE_URL")
        relationMode = "prisma"
        extensions = [vector]
      }
</pre>

**Reply by KyungmoKu** [reply]
While using Supabase & Prisma,
adding previewFeatures and extensions somehow caused migration drift.

Solved by deleting these and adding pg-vector extension manually on supabase.

**Comment by chrbala** [comment]
Is there a way to set this up for the [bit data type](https://github.com/pgvector/pgvector?tab=readme-ov-file#binary-vectors)? I tried and when I attempted to set the field with array of 0 and 1s via `prisma.$executeRaw`, I got `cannot cast type bigint[] to bit`.
